internal/querycoordv2/meta/replica_manager.go

```diff
@@ -283,20 +271,88 @@ func (m *ReplicaManager) GetByResourceGroup(rgName string) []*Replica {
 	return ret
 }
 
-func (m *ReplicaManager) AddNode(replicaID typeutil.UniqueID, nodes ...typeutil.UniqueID) error {
+// RecoverNodesInCollection recovers all nodes in collection with latest resource group.
+// Promise a node will be only assigned to one replica in same collection at same time.
+// 1. Move the nodes to outbound if they are not in related resource group.
+// 2. Add new incoming nodes into the replica if they are not in-used by other replicas of same collection.
+// 3. replicas in same resource group will shared the nodes in resource group fairly.
+func (m *ReplicaManager) RecoverNodesInCollection(collectionID typeutil.UniqueID, rgs map[string]typeutil.UniqueSet) error {
+	if err := m.validateResourceGroups(rgs); err != nil {
+		return err
+	}
+
 	m.rwmutex.Lock()
 	defer m.rwmutex.Unlock()
 
-	replica, ok := m.replicas[replicaID]
+	// create a helper to do the recover.
+	helper, err := m.getReplicaAssignmentHelper(collectionID, rgs)
+	if err != nil {
+		return err
+	}
+
+	modifiedReplicas := make([]*Replica, 0)
+	// recover node by resource group.
+	helper.RangeOverResourceGroup(func(replicaHelper *replicaAssignmentHelper) {
+		replicaHelper.RangeOverReplicas(func(assignment *replicaAssignmentInfo) {
+			outboundNodes := assignment.GetNewOutboundNodes()
+			recoverableNodes, incomingNodeCount := assignment.GetRecoverNodesAndIncomingNodeCount()
```

Comment:
recomand to use `replicaAssignmentHelper` to generate a immutable assignment, then apply to replica_manager. change assignment result during `getFunc` is weird

Reply:
collectionAssignmentHelper itself is a new type that has no share memory with orignal replicas in replica manager.
The only mutable part is `AllocateInComingNode`, others are immutable.